validations/src/eligibility/current.rs

```diff
@@ -147,37 +151,41 @@ where
             Err(e) => {
                 // Early exit if the stake key does not exist
                 return match e {
-                    StakesError::EntryNotFound { .. } => Ok(IneligibilityReason::NotStaking.into()),
+                    StakesError::EntryNotFound { .. } => {
+                        Ok((IneligibilityReason::NotStaking.into(), Hash::default()))
+                    }
                     e => Err(e),
                 };
             }
         };
 
-        let mut rank = self.rank(Capability::Mining, epoch);
-
-        // Requirement no. 3 from the WIP:
-        //  "the mining power of the block proposer is greater than `max_power / rf`"
-        // (This goes before no. 2 because it is cheaper, computation-wise, and we want validations to exit ASAP)
-        // TODO: verify if defaulting to 0 makes sense
-        let (_, max_power) = rank.next().unwrap_or_default();
-        let threshold = max_power / Power::from(MINING_REPLICATION_FACTOR as u64);
-        if power <= threshold {
-            return Ok(IneligibilityReason::InsufficientPower.into());
-        }
-
         // Requirement no. 2 from the WIP:
         //  "the mining power of the block proposer is in the `rf / stakers`th quantile among the mining powers of all
         //  the stakers"
-        let stakers = self.stakes_count();
-        let quantile = stakers / MINING_REPLICATION_FACTOR;
         // TODO: verify if defaulting to 0 makes sense
-        let (_, threshold) = rank.nth(quantile).unwrap_or_default();
-        if power <= threshold {
-            return Ok(IneligibilityReason::InsufficientPower.into());
+        let mut rank = self.rank(Capability::Mining, epoch);
+        let (_, max_power) = rank.next().unwrap_or_default();
+        let (_, threshold) = rank.nth(MINING_REPLICATION_FACTOR - 2).unwrap_or_default();
```

Comment:
Shouldn't it be `rank.nth(MINING_REPLICATION_FACTOR -1)` instead?

Reply:
No, because `rank.next()` consumes element 0 of the iterator and you want the element up to `MINING_REPLICATION_FACTOR - 1`. If `rf` is 4, line 167 consumes element 0 and `nth(2)` returns and consumes up to the `nth` element starting from 0.